Improvements in resource conflict detection related to slots

martien-de-jong opened the pull request:
This started out by the realization that XM is a separate slot from X and M, and that the only mechanism that prevents them from occupying the same bundle is the format table. 

ResourceCycle will make a quick check whether there's overlap in the slot bits, but will miss an X XM overlap.

The idea is that we can go through the complete format list and find the implications from XM to X and M (All formats containing XM don't allow X or M)

First, we have simplified the generated tables by removing the target-specific SlotInfo and SlotKind classes. That removes the need for MCFormats.h to include all generated format tables. This aligns with the tablegen idiom that we have universal data type whose data is supplied as an opaque bunch of untyped data.

Trying to use the denser 'ConflictBits' in the check pointed out a problem, which is the nop slot as used by AIE2 to disambiguate  the two stand-alone formats for XM, on in 48 bits and one in 64 bits. The latter one is only used for bundle padding. That nop slot only occurs in the 16 nop format and the 64 bit XM format, and hence is implied by A, B, S, X and M slots. 
We added the notion of an artificial slot to be able to exclude it from the conflict bits.

**Status**

- Clean-up of slot and format interfaces
- Conflict bits generated and used.
- NFC for  existing unit tests
- Use a O(1) order lookup for format availability
- Strip some unused data from the field description table in AIEMCFormats.h

Commit:
[AIE] SlotID counter can be local

Commit:
[AIE] Kind/getSlotKind unused

Commit:
[AIE] global UNKNOWN_SLOT declaration

Commit:
[AIE] Express everything as MCSlotKind/MCSlotInfo

martien-de-jong commented on llvm/utils/TableGen/CodeGenFormat.cpp:
```diff
@@ -974,8 +974,6 @@ void TGFieldLayout::resolveFieldsGlobalOffsets() {
   GlobalOffsets.RightOffset = GlobalOffsets.LeftOffset + this->getSize() - 1;
 }
 
-unsigned TGTargetSlots::SlotID = 0;
```
nit copyright change here.

martien-de-jong commented on llvm/lib/Target/AIE/MCTargetDesc/AIEMCFormats.h:
```diff
@@ -26,6 +26,8 @@
 namespace llvm {
 
+static const int SLOT_UNKNOWN = -1;
+
```
nit cpyrgt change here

martien-de-jong commented on llvm/utils/TableGen/CodeGenFormat.h:
```diff
@@ -404,9 +404,8 @@ class TGTargetSlots {
-
-  const_iterator cend() const { return Slots.end(); }
+  const_iterator begin() const { return Slots.begin(); }
+  const_iterator end() const { return Slots.end(); }
```
There's probably a reason, e.g. being able to use range-based syntax.

martien-de-jong commented on llvm/lib/Target/AIE/MCTargetDesc/AIEMCFormats.h:
```diff
@@ -75,21 +75,31 @@ class MCSlotInfo {
+      : SlotName(SlotName), Size(Size), SlotOccupancy(Bits), ConflictBits(Bits),
+        NopOpc(NopOpc) {}
+  constexpr MCSlotInfo(int Kind, const char *SlotName, unsigned Size,
+                       SlotBits Bits, SlotBits Conflicts, unsigned NopOpc)
```
nit: it looks as we could lose Kind plus related constructors.

martien-de-jong commented on llvm/utils/TableGen/CodeGenFormat.cpp:
```diff
@@ -220,7 +216,6 @@ void CodeGenFormat::run(raw_ostream &o) {
 
     for (TGInstrLayout &Inst : InstFormats)
       if (Inst.isPacketFormat()) {
-        Inst.computeSlotSet();
```
We've moved this out to a second pass, since we first should process all formats before we can compute the conflict bits.

Commit:
[AIE][CodeGenFormat] lose a lazy constructor

mludevid commented on llvm/utils/TableGen/CodeGenFormat.cpp:
```diff
@@ -1171,16 +1171,15 @@ void TGTargetSlots::emitTargetSlotClass(raw_ostream &o) const {
 
   o << "class " << TargetClassName << " : public MC" << GenSlotInfoName << "\n"
     << "{\n"
-    << "  const " << TargetEnumName << " Kind;\n"
```
Please also remove the `const std::string TargetEnumName = ...`.

martien-de-jong replied:
Actually, the entire function disappears later on.

mludevid commented on llvm/utils/TableGen/CodeGenFormat.cpp:
```diff
@@ -631,15 +630,14 @@ void TGInstrLayout::emitFlatTree(ConstTable &FieldsHierarchy,
     else
-      FieldsHierarchy << DefaultSlotInfo->second.getEnumerationString();
-    FieldsHierarchy << ") }";
+      FieldsHierarchy << "SLOT_UNKNOWN";
```
Can we create a global `static const` for `"SLOT_UNKONWN"` as well? It is used in a few places...

martien-de-jong replied:
I think we have one in AIEMCFormats.h. This is its use.

martien-de-jong commented on llvm/utils/TableGen/CodeGenFormat.cpp:
```diff
@@ -251,52 +249,91 @@ void CodeGenFormat::run(raw_ostream &o) {
+    }
+    Slot.setConflictBits(ThisSlot | Excluded);
+  }
+}
```
Ehh, this got lost in rebase. This is NOT a cosmetic detail.

Commit:
[AIE][CodeGenFormat] Compute ConflictBits using all given formats

Separate SlotBits from ConflictBits

Commit:
[AIE] Use ConflictBits in bundle conflict test

martien-de-jong commented on llvm/lib/Target/AIE/AIEBundle.h:
```diff
@@ -91,9 +91,15 @@ template <class I> class Bundle {
+
+    auto *SlotInfo = FormatInterface->getSlotInfo(Slot);
+    // ConflictBits is a fast predictor of missing formats
+    SlotBits ConflictBits = SlotInfo->getConflictSet();
```
Using ConflictBits here is easy. It's a pity that this is not the biggest hotspot. The more frequently used conflict interfaces in AIEHazardRecognizer are more of a problem.

Commit:
[AIE] Use ConflictBits in HazardRecognizer conflict interfaces

martien-de-jong commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -125,6 +125,7 @@ FuncUnitWrapper &FuncUnitWrapper::operator|=(const FuncUnitWrapper &Other) {
 bool FuncUnitWrapper::conflict(const FuncUnitWrapper &Other) const {
   if ((Slots & Other.Slots) != 0 || (MemoryBanks & Other.MemoryBanks) != 0 ||
       (MemObjectsBits & Other.MemObjectsBits) != 0 ||
+      (Conflicts & Other.Slots) != 0 || (Slots & Other.Conflicts) != 0 ||
```
This was the easiest way. We misuse FuncUnitWrapper to hold the auxiliary data in Conflicts, which is a straight copy of the SlotInfo field. Conflicts isn't part of the scoreboard state, just of the incoming new FuncUnitWrapper.
-- well, it will get merged into the scoreboard after all, it keeps the merge operation associative.

martien-de-jong commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -482,6 +483,15 @@ static SlotBits getSlotSet(const MCInstrDesc &Desc,
+  if (SlotKind != MCSlotKind())
+    return Formats.getSlotInfo(SlotKind)->getConflictSet();
+
+  return 0;
```
If we can't retrieve conflict bits, we fall back to the existing slot test.

martien-de-jong commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -540,25 +551,26 @@ bool AIEHazardRecognizer::checkConflict(
   // Verify memory bank and shared object hazards
   if (!MemoryAccessCycles.empty()) {
-    FuncUnitWrapper MemoryAccessCycle(/*SlotSet=*/0, MemoryBanks,
+    FuncUnitWrapper MemoryAccessCycle(/*SlotSet=*/0, 0, MemoryBanks,
```
The danger of weakly typed default parameters. Without this change, it compiled, but all memory bank tests broke.

martien-de-jong commented on llvm/lib/Target/AIE/AIEHazardRecognizer.h:
```diff
@@ -72,6 +72,9 @@ class FuncUnitWrapper {
   SlotBits Slots = 0;
+  /// Conflicts are just for speeding up conflict detection. They may be present
+  /// in cycles to be merged, but they will not be merged into the scoreboard.
+  SlotBits Conflicts = 0;
```
TODO: I need to clear these bits. Moreover, we will merge the bits, since we can't know whether we are incrementally building a cycle to add to the scoreboard, or merging into the scoreboard -- both are done using `operator |`.

Commit:
[AIE][CodeGenFormat] Cosmetics

Conflict resolution

Commit:
[AIE] Remove obsolete interface

All accesses to the packetformats run through the more complete
getFormatInterface

Commit:
[AIE] Improved format availability check

Commit:
[AIE] Strip out unused fields from the FieldsHierarchyTable

Commit:
[AIE] SImplify seraching for formats by size

Commit:
[AIE] Prevent double checks. use llvm::find_if

Commit:
fix CodeGenFormat tests

martien-de-jong commented:
Some of the cleanup commits have been put into a separate PR. The rest will resume in a new PR.
